File: frontend/src/engine/math/aabb.ts
/** Axis-aligned bounding box. Mirrors backend/app/domain/math2d/aabb.py. */

import type { Vec2 } from "./vec2";

export interface Aabb {
  readonly minX: number;
  readonly minY: number;
  readonly maxX: number;
  readonly maxY: number;
}

export function aabb(minX: number, minY: number, maxX: number, maxY: number): Aabb {
  if (minX > maxX || minY > maxY) {
    throw new Error(`invalid AABB: min (${minX}, ${minY}) exceeds max (${maxX}, ${maxY})`);
  }
  return { minX, minY, maxX, maxY };
}

export function aabbFromPoints(points: readonly Vec2[]): Aabb {
  if (points.length === 0) {
    throw new Error("cannot build an AABB from zero points");
  }
  let minX = points[0].x;
  let minY = points[0].y;
  let maxX = points[0].x;
  let maxY = points[0].y;
  for (const point of points.slice(1)) {
    minX = Math.min(minX, point.x);
    minY = Math.min(minY, point.y);
    maxX = Math.max(maxX, point.x);
    maxY = Math.max(maxY, point.y);
  }
  return { minX, minY, maxX, maxY };
}

export function aabbWidth(box: Aabb): number {
  return box.maxX - box.minX;
}

export function aabbHeight(box: Aabb): number {
  return box.maxY - box.minY;
}

export function aabbCenter(box: Aabb): Vec2 {
  return { x: (box.minX + box.maxX) / 2, y: (box.minY + box.maxY) / 2 };
}

export function aabbContainsPoint(box: Aabb, point: Vec2): boolean {
  return (
    box.minX <= point.x && point.x <= box.maxX && box.minY <= point.y && point.y <= box.maxY
  );
}

export function aabbIntersects(a: Aabb, b: Aabb): boolean {
  return a.minX <= b.maxX && b.minX <= a.maxX && a.minY <= b.maxY && b.minY <= a.maxY;
}

export function aabbUnion(a: Aabb, b: Aabb): Aabb {
  return {
    minX: Math.min(a.minX, b.minX),
    minY: Math.min(a.minY, b.minY),
    maxX: Math.max(a.maxX, b.maxX),
    maxY: Math.max(a.maxY, b.maxY)
  };
}

export function aabbExpanded(box: Aabb, margin: number): Aabb {
  return {
    minX: box.minX - margin,
    minY: box.minY - margin,
    maxX: box.maxX + margin,
    maxY: box.maxY + margin
  };
}

File: backend/app/domain/math2d/aabb.py
"""Axis-aligned bounding box."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass

from app.domain.math2d.vec2 import Vec2


@dataclass(frozen=True, slots=True)
class Aabb:
    min_x: float
    min_y: float
    max_x: float
    max_y: float

    def __post_init__(self) -> None:
        if self.min_x > self.max_x or self.min_y > self.max_y:
            raise ValueError(
                f"invalid AABB: min ({self.min_x}, {self.min_y}) "
                f"exceeds max ({self.max_x}, {self.max_y})"
            )

    @staticmethod
    def from_points(points: Iterable[Vec2]) -> Aabb:
        iterator = iter(points)
        try:
            first = next(iterator)
        except StopIteration:
            raise ValueError("cannot build an AABB from zero points") from None
        min_x = max_x = first.x
        min_y = max_y = first.y
        for point in iterator:
            min_x = min(min_x, point.x)
            min_y = min(min_y, point.y)
            max_x = max(max_x, point.x)
            max_y = max(max_y, point.y)
        return Aabb(min_x, min_y, max_x, max_y)

    def width(self) -> float:
        return self.max_x - self.min_x

    def height(self) -> float:
        return self.max_y - self.min_y

    def center(self) -> Vec2:
        return Vec2((self.min_x + self.max_x) / 2.0, (self.min_y + self.max_y) / 2.0)

    def contains_point(self, point: Vec2) -> bool:
        return self.min_x <= point.x <= self.max_x and self.min_y <= point.y <= self.max_y

    def intersects(self, other: Aabb) -> bool:
        return (
            self.min_x <= other.max_x
            and other.min_x <= self.max_x
            and self.min_y <= other.max_y
            and other.min_y <= self.max_y
        )

    def union(self, other: Aabb) -> Aabb:
        return Aabb(
            min(self.min_x, other.min_x),
            min(self.min_y, other.min_y),
            max(self.max_x, other.max_x),
            max(self.max_y, other.max_y),
        )

    def expanded(self, margin: float) -> Aabb:
        return Aabb(
            self.min_x - margin,
            self.min_y - margin,
            self.max_x + margin,
            self.max_y + margin,
        )
